File: src-tauri/src/commands/annotations.rs
use crate::db::migrations::DbPool;
use crate::db::models::{Highlight, MarginNote};
use rusqlite::Connection;
use std::time::SystemTime;
use uuid::Uuid;

fn now_millis() -> i64 {
    SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

fn touch_document(conn: &Connection, document_id: &str) -> Result<(), String> {
    conn.execute(
        "UPDATE documents SET last_opened_at = ?1 WHERE id = ?2",
        rusqlite::params![now_millis(), document_id],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

fn document_id_for_highlight(conn: &Connection, highlight_id: &str) -> Result<String, String> {
    conn.query_row(
        "SELECT document_id FROM highlights WHERE id = ?1",
        rusqlite::params![highlight_id],
        |row| row.get(0),
    )
    .map_err(|e| e.to_string())
}

fn document_id_for_margin_note(conn: &Connection, note_id: &str) -> Result<String, String> {
    conn.query_row(
        "SELECT h.document_id FROM margin_notes mn JOIN highlights h ON mn.highlight_id = h.id WHERE mn.id = ?1",
        rusqlite::params![note_id],
        |row| row.get(0),
    )
    .map_err(|e| e.to_string())
}

// === Inner functions (testable with &Connection) ===

fn insert_highlight(
    conn: &Connection,
    id: &str,
    document_id: &str,
    color: &str,
    text_content: &str,
    from_pos: i64,
    to_pos: i64,
    prefix_context: Option<&str>,
    suffix_context: Option<&str>,
    now: i64,
) -> Result<(), String> {
    conn.execute(
        "INSERT INTO highlights
            (id, document_id, color, text_content, from_pos, to_pos,
             prefix_context, suffix_context, created_at, updated_at)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10)",
        rusqlite::params![id, document_id, color, text_content, from_pos, to_pos, prefix_context, suffix_context, now, now],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

fn fetch_highlights(conn: &Connection, document_id: &str) -> Result<Vec<Highlight>, String> {
    let mut stmt = conn
        .prepare(
            "SELECT id, document_id, color, text_content, from_pos, to_pos,
                    prefix_context, suffix_context, created_at, updated_at
             FROM highlights
             WHERE document_id = ?1
             ORDER BY from_pos",
        )
        .map_err(|e| e.to_string())?;

    let results = stmt
        .query_map([document_id], |row| Highlight::from_row(row))
        .map_err(|e| e.to_string())?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| e.to_string());
    results
}

fn set_highlight_color(conn: &Connection, id: &str, color: &str, now: i64) -> Result<(), String> {
    conn.execute(
        "UPDATE highlights SET color = ?1, updated_at = ?2 WHERE id = ?3",
        rusqlite::params![color, now, id],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

fn remove_highlight(conn: &Connection, id: &str) -> Result<(), String> {
    conn.execute("DELETE FROM highlights WHERE id = ?1", rusqlite::params![id])
        .map_err(|e| e.to_string())?;
    Ok(())
}

fn insert_margin_note(
    conn: &Connection,
    id: &str,
    highlight_id: &str,
    content: &str,
    now: i64,
) -> Result<(), String> {
    conn.execute(
        "INSERT INTO margin_notes (id, highlight_id, content, created_at, updated_at)
         VALUES (?1, ?2, ?3, ?4, ?5)",
        rusqlite::params![id, highlight_id, content, now, now],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

fn fetch_margin_notes(conn: &Connection, document_id: &str) -> Result<Vec<MarginNote>, String> {
    let mut stmt = conn
        .prepare(
            "SELECT mn.id, mn.highlight_id, mn.content, mn.created_at, mn.updated_at
             FROM margin_notes mn
             JOIN highlights h ON mn.highlight_id = h.id
             WHERE h.document_id = ?1
             ORDER BY h.from_pos",
        )
        .map_err(|e| e.to_string())?;

    let results = stmt
        .query_map([document_id], |row| MarginNote::from_row(row))
        .map_err(|e| e.to_string())?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| e.to_string());
    results
}

fn set_margin_note_content(conn: &Connection, id: &str, content: &str, now: i64) -> Result<(), String> {
    conn.execute(
        "UPDATE margin_notes SET content = ?1, updated_at = ?2 WHERE id = ?3",
        rusqlite::params![content, now, id],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

fn remove_margin_note(conn: &Connection, id: &str) -> Result<(), String> {
    conn.execute(
        "DELETE FROM margin_notes WHERE id = ?1",
        rusqlite::params![id],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

fn remove_all_highlights_for_document(conn: &Connection, document_id: &str) -> Result<usize, String> {
    conn.execute(
        "DELETE FROM highlights WHERE document_id = ?1",
        rusqlite::params![document_id],
    )
    .map_err(|e| e.to_string())
}

// === Tauri command handlers ===

#[tauri::command]
pub async fn create_highlight(
    state: tauri::State<'_, DbPool>,
    document_id: String,
    color: String,
    text_content: String,
    from_pos: i64,
    to_pos: i64,
    prefix_context: Option<String>,
    suffix_context: Option<String>,
) -> Result<Highlight, String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());
    let id = Uuid::new_v4().to_string();
    let now = now_millis();

    insert_highlight(
        &conn, &id, &document_id, &color, &text_content,
        from_pos, to_pos,
        prefix_context.as_deref(), suffix_context.as_deref(),
        now,
    )?;

    touch_document(&conn, &document_id)?;

    Ok(Highlight {
        id,
        document_id,
        color,
        text_content,
        from_pos,
        to_pos,
        prefix_context,
        suffix_context,
        created_at: now,
        updated_at: now,
    })
}

#[tauri::command]
pub async fn get_highlights(state: tauri::State<'_, DbPool>, document_id: String) -> Result<Vec<Highlight>, String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());
    fetch_highlights(&conn, &document_id)
}

#[tauri::command]
pub async fn update_highlight_color(state: tauri::State<'_, DbPool>, id: String, color: String) -> Result<(), String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());
    let now = now_millis();

    set_highlight_color(&conn, &id, &color, now)?;

    let doc_id = document_id_for_highlight(&conn, &id)?;
    touch_document(&conn, &doc_id)?;

    Ok(())
}

#[tauri::command]
pub async fn delete_highlight(state: tauri::State<'_, DbPool>, id: String) -> Result<(), String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());

    let doc_id = document_id_for_highlight(&conn, &id)?;
    remove_highlight(&conn, &id)?;
    touch_document(&conn, &doc_id)?;

    Ok(())
}

#[tauri::command]
pub async fn create_margin_note(
    state: tauri::State<'_, DbPool>,
    highlight_id: String,
    content: String,
) -> Result<MarginNote, String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());
    let id = Uuid::new_v4().to_string();
    let now = now_millis();

    insert_margin_note(&conn, &id, &highlight_id, &content, now)?;

    let doc_id = document_id_for_highlight(&conn, &highlight_id)?;
    touch_document(&conn, &doc_id)?;

    Ok(MarginNote {
        id,
        highlight_id,
        content,
        created_at: now,
        updated_at: now,
    })
}

#[tauri::command]
pub async fn get_margin_notes(state: tauri::State<'_, DbPool>, document_id: String) -> Result<Vec<MarginNote>, String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());
    fetch_margin_notes(&conn, &document_id)
}

#[tauri::command]
pub async fn update_margin_note(state: tauri::State<'_, DbPool>, id: String, content: String) -> Result<(), String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());
    let now = now_millis();

    let doc_id = document_id_for_margin_note(&conn, &id)?;
    set_margin_note_content(&conn, &id, &content, now)?;
    touch_document(&conn, &doc_id)?;

    Ok(())
}

#[tauri::command]
pub async fn delete_margin_note(state: tauri::State<'_, DbPool>, id: String) -> Result<(), String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());

    let doc_id = document_id_for_margin_note(&conn, &id)?;
    remove_margin_note(&conn, &id)?;
    touch_document(&conn, &doc_id)?;

    Ok(())
}

#[tauri::command]
pub async fn delete_all_highlights_for_document(
    state: tauri::State<'_, DbPool>,
    document_id: String,
) -> Result<usize, String> {
    let conn = state.0.lock().unwrap_or_else(|e| e.into_inner());
    remove_all_highlights_for_document(&conn, &document_id)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn schema_sql() -> &'static str {
        "PRAGMA foreign_keys=ON;
         CREATE TABLE documents (
             id TEXT PRIMARY KEY,
             source TEXT NOT NULL,
             file_path TEXT,
             keep_local_id TEXT,
             title TEXT,
             author TEXT,
             url TEXT,
             word_count INTEGER DEFAULT 0,
             last_opened_at INTEGER NOT NULL,
             created_at INTEGER NOT NULL,
             UNIQUE(file_path),
             UNIQUE(keep_local_id)
         );
         CREATE TABLE highlights (
             id TEXT PRIMARY KEY,
             document_id TEXT NOT NULL REFERENCES documents(id) ON DELETE CASCADE,
             color TEXT NOT NULL DEFAULT 'yellow',
             text_content TEXT NOT NULL,
             from_pos INTEGER NOT NULL,
             to_pos INTEGER NOT NULL,
             prefix_context TEXT,
             suffix_context TEXT,
             created_at INTEGER NOT NULL,
             updated_at INTEGER NOT NULL
         );
         CREATE INDEX idx_highlights_document ON highlights(document_id);
         CREATE TABLE margin_notes (
             id TEXT PRIMARY KEY,
             highlight_id TEXT NOT NULL REFERENCES highlights(id) ON DELETE CASCADE,
             content TEXT NOT NULL,
             created_at INTEGER NOT NULL,
             updated_at INTEGER NOT NULL
         );
         CREATE INDEX idx_margin_notes_highlight ON margin_notes(highlight_id);"
    }

    fn setup_db() -> Connection {
        let conn = Connection::open_in_memory().unwrap();
        conn.execute_batch(schema_sql()).unwrap();
        conn
    }

    fn insert_doc(conn: &Connection, id: &str) {
        conn.execute(
            "INSERT INTO documents (id, source, title, last_opened_at, created_at)
             VALUES (?1, 'file', 'Test Doc', 1000, 1000)",
            rusqlite::params![id],
        )
        .unwrap();
    }

    fn highlight_count(conn: &Connection) -> i64 {
        conn.query_row("SELECT COUNT(*) FROM highlights", [], |r| r.get(0)).unwrap()
    }

    fn note_count(conn: &Connection) -> i64 {
        conn.query_row("SELECT COUNT(*) FROM margin_notes", [], |r| r.get(0)).unwrap()
    }

    // === Highlight tests ===

    #[test]
    fn insert_and_fetch_highlights() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");

        insert_highlight(&conn, "h1", "doc1", "yellow", "hello world", 0, 11, Some("pre"), Some("suf"), 1000).unwrap();
        insert_highlight(&conn, "h2", "doc1", "green", "second", 20, 26, None, None, 1001).unwrap();

        let highlights = fetch_highlights(&conn, "doc1").unwrap();
        assert_eq!(highlights.len(), 2);
        assert_eq!(highlights[0].id, "h1");
        assert_eq!(highlights[0].text_content, "hello world");
        assert_eq!(highlights[0].prefix_context.as_deref(), Some("pre"));
        assert_eq!(highlights[1].id, "h2");
        assert_eq!(highlights[1].color, "green");
    }

    #[test]
    fn fetch_highlights_ordered_by_from_pos() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");

        // Insert out of order
        insert_highlight(&conn, "h2", "doc1", "yellow", "later", 50, 55, None, None, 1000).unwrap();
        insert_highlight(&conn, "h1", "doc1", "yellow", "earlier", 10, 17, None, None, 1001).unwrap();

        let highlights = fetch_highlights(&conn, "doc1").unwrap();
        assert_eq!(highlights[0].from_pos, 10);
        assert_eq!(highlights[1].from_pos, 50);
    }

    #[test]
    fn fetch_highlights_scoped_to_document() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_doc(&conn, "doc2");

        insert_highlight(&conn, "h1", "doc1", "yellow", "in doc1", 0, 7, None, None, 1000).unwrap();
        insert_highlight(&conn, "h2", "doc2", "yellow", "in doc2", 0, 7, None, None, 1000).unwrap();

        let highlights = fetch_highlights(&conn, "doc1").unwrap();
        assert_eq!(highlights.len(), 1);
        assert_eq!(highlights[0].document_id, "doc1");
    }

    #[test]
    fn update_highlight_color_changes_color_and_timestamp() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();

        set_highlight_color(&conn, "h1", "green", 2000).unwrap();

        let highlights = fetch_highlights(&conn, "doc1").unwrap();
        assert_eq!(highlights[0].color, "green");
        assert_eq!(highlights[0].updated_at, 2000);
        assert_eq!(highlights[0].created_at, 1000); // unchanged
    }

    #[test]
    fn delete_highlight_removes_it() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();
        assert_eq!(highlight_count(&conn), 1);

        remove_highlight(&conn, "h1").unwrap();
        assert_eq!(highlight_count(&conn), 0);
    }

    #[test]
    fn delete_highlight_cascades_to_margin_notes() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();
        insert_margin_note(&conn, "n1", "h1", "my note", 1000).unwrap();
        assert_eq!(note_count(&conn), 1);

        remove_highlight(&conn, "h1").unwrap();
        assert_eq!(note_count(&conn), 0); // cascade
    }

    #[test]
    fn delete_document_cascades_to_highlights_and_notes() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();
        insert_margin_note(&conn, "n1", "h1", "my note", 1000).unwrap();

        conn.execute("DELETE FROM documents WHERE id = 'doc1'", []).unwrap();
        assert_eq!(highlight_count(&conn), 0);
        assert_eq!(note_count(&conn), 0);
    }

    // === Margin Note tests ===

    #[test]
    fn insert_and_fetch_margin_notes() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();

        insert_margin_note(&conn, "n1", "h1", "first note", 1000).unwrap();
        insert_margin_note(&conn, "n2", "h1", "second note", 1001).unwrap();

        let notes = fetch_margin_notes(&conn, "doc1").unwrap();
        assert_eq!(notes.len(), 2);
        assert_eq!(notes[0].content, "first note");
        assert_eq!(notes[0].highlight_id, "h1");
    }

    #[test]
    fn fetch_margin_notes_ordered_by_highlight_position() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h2", "doc1", "yellow", "later", 50, 55, None, None, 1000).unwrap();
        insert_highlight(&conn, "h1", "doc1", "yellow", "earlier", 10, 17, None, None, 1000).unwrap();

        insert_margin_note(&conn, "n2", "h2", "note on later", 1000).unwrap();
        insert_margin_note(&conn, "n1", "h1", "note on earlier", 1001).unwrap();

        let notes = fetch_margin_notes(&conn, "doc1").unwrap();
        assert_eq!(notes[0].content, "note on earlier");
        assert_eq!(notes[1].content, "note on later");
    }

    #[test]
    fn update_margin_note_changes_content_and_timestamp() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();
        insert_margin_note(&conn, "n1", "h1", "original", 1000).unwrap();

        set_margin_note_content(&conn, "n1", "updated", 2000).unwrap();

        let notes = fetch_margin_notes(&conn, "doc1").unwrap();
        assert_eq!(notes[0].content, "updated");
        assert_eq!(notes[0].updated_at, 2000);
        assert_eq!(notes[0].created_at, 1000);
    }

    #[test]
    fn delete_margin_note_removes_it() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();
        insert_margin_note(&conn, "n1", "h1", "note", 1000).unwrap();
        assert_eq!(note_count(&conn), 1);

        remove_margin_note(&conn, "n1").unwrap();
        assert_eq!(note_count(&conn), 0);
    }

    #[test]
    fn document_id_for_highlight_returns_correct_doc() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();

        assert_eq!(document_id_for_highlight(&conn, "h1").unwrap(), "doc1");
    }

    #[test]
    fn document_id_for_margin_note_returns_correct_doc() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_highlight(&conn, "h1", "doc1", "yellow", "text", 0, 4, None, None, 1000).unwrap();
        insert_margin_note(&conn, "n1", "h1", "note", 1000).unwrap();

        assert_eq!(document_id_for_margin_note(&conn, "n1").unwrap(), "doc1");
    }

    #[test]
    fn empty_document_returns_no_highlights() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");

        let highlights = fetch_highlights(&conn, "doc1").unwrap();
        assert!(highlights.is_empty());
    }

    #[test]
    fn empty_document_returns_no_margin_notes() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");

        let notes = fetch_margin_notes(&conn, "doc1").unwrap();
        assert!(notes.is_empty());
    }

    // === Batch delete tests ===

    #[test]
    fn remove_all_highlights_for_document_clears_all() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");
        insert_doc(&conn, "doc2");

        // doc1: 2 highlights, 1 with a note
        insert_highlight(&conn, "h1", "doc1", "yellow", "text1", 0, 5, None, None, 1000).unwrap();
        insert_highlight(&conn, "h2", "doc1", "green", "text2", 10, 15, None, None, 1000).unwrap();
        insert_margin_note(&conn, "n1", "h1", "note on h1", 1000).unwrap();

        // doc2: 1 highlight with a note (should be untouched)
        insert_highlight(&conn, "h3", "doc2", "blue", "text3", 0, 5, None, None, 1000).unwrap();
        insert_margin_note(&conn, "n2", "h3", "note on h3", 1000).unwrap();

        let deleted = remove_all_highlights_for_document(&conn, "doc1").unwrap();
        assert_eq!(deleted, 2);

        // doc1 is empty
        assert!(fetch_highlights(&conn, "doc1").unwrap().is_empty());
        assert!(fetch_margin_notes(&conn, "doc1").unwrap().is_empty());

        // doc2 untouched
        assert_eq!(fetch_highlights(&conn, "doc2").unwrap().len(), 1);
        assert_eq!(fetch_margin_notes(&conn, "doc2").unwrap().len(), 1);
    }

    #[test]
    fn remove_all_highlights_for_document_empty_is_noop() {
        let conn = setup_db();
        insert_doc(&conn, "doc1");

        let deleted = remove_all_highlights_for_document(&conn, "doc1").unwrap();
        assert_eq!(deleted, 0);
    }
}

File: mcp/src/tools/annotations.ts
import type Database from "better-sqlite3";
import { randomUUID } from "crypto";
import { nowMillis, touchDocument } from "../db.js";
import { readDocument, getDocument } from "./documents.js";

export interface HighlightRecord {
  id: string;
  document_id: string;
  color: string;
  text_content: string;
  from_pos: number;
  to_pos: number;
  prefix_context: string | null;
  suffix_context: string | null;
  created_at: number;
  updated_at: number;
}

export interface MarginNoteRecord {
  id: string;
  highlight_id: string;
  content: string;
  created_at: number;
  updated_at: number;
}

export interface AnnotationEntry {
  highlight: HighlightRecord;
  notes: MarginNoteRecord[];
}

const ALLOWED_COLORS = [
  "yellow",
  "green",
  "blue",
  "pink",
  "purple",
  "orange",
] as const;

export function getAnnotations(
  db: Database.Database,
  documentId: string,
): AnnotationEntry[] {
  const highlights = db
    .prepare(
      `SELECT id, document_id, color, text_content, from_pos, to_pos,
              prefix_context, suffix_context, created_at, updated_at
       FROM highlights
       WHERE document_id = ?
       ORDER BY from_pos`,
    )
    .all(documentId) as HighlightRecord[];

  if (highlights.length === 0) return [];

  const allNotes = db
    .prepare(
      `SELECT mn.id, mn.highlight_id, mn.content, mn.created_at, mn.updated_at
       FROM margin_notes mn
       JOIN highlights h ON mn.highlight_id = h.id
       WHERE h.document_id = ?
       ORDER BY h.from_pos, mn.created_at`,
    )
    .all(documentId) as MarginNoteRecord[];

  // Group notes by highlight_id
  const notesByHighlight = new Map<string, MarginNoteRecord[]>();
  for (const note of allNotes) {
    const existing = notesByHighlight.get(note.highlight_id);
    if (existing) {
      existing.push(note);
    } else {
      notesByHighlight.set(note.highlight_id, [note]);
    }
  }

  return highlights.map((highlight) => ({
    highlight,
    notes: notesByHighlight.get(highlight.id) ?? [],
  }));
}

export function createHighlight(
  db: Database.Database,
  params: {
    document_id: string;
    color: string;
    text_content: string;
    from_pos: number;
    to_pos: number;
    prefix_context?: string;
    suffix_context?: string;
  },
): HighlightRecord | { error: string } {
  if (!ALLOWED_COLORS.includes(params.color as (typeof ALLOWED_COLORS)[number])) {
    return {
      error: `Invalid color "${params.color}". Allowed: ${ALLOWED_COLORS.join(", ")}`,
    };
  }

  if (
    !Number.isInteger(params.from_pos) ||
    !Number.isInteger(params.to_pos) ||
    params.from_pos < 0 ||
    params.from_pos >= params.to_pos
  ) {
    return {
      error: `Invalid position range: from_pos (${params.from_pos}) must be a non-negative integer less than to_pos (${params.to_pos}).`,
    };
  }

  // Verify document exists
  const doc = db
    .prepare("SELECT id FROM documents WHERE id = ?")
    .get(params.document_id);
  if (!doc) {
    return { error: `Document not found: ${params.document_id}` };
  }

  const id = randomUUID();
  const now = nowMillis();

  db.prepare(
    `INSERT INTO highlights
       (id, document_id, color, text_content, from_pos, to_pos,
        prefix_context, suffix_context, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
  ).run(
    id,
    params.document_id,
    params.color,
    params.text_content,
    params.from_pos,
    params.to_pos,
    params.prefix_context ?? null,
    params.suffix_context ?? null,
    now,
    now,
  );

  touchDocument(db, params.document_id);

  return {
    id,
    document_id: params.document_id,
    color: params.color,
    text_content: params.text_content,
    from_pos: params.from_pos,
    to_pos: params.to_pos,
    prefix_context: params.prefix_context ?? null,
    suffix_context: params.suffix_context ?? null,
    created_at: now,
    updated_at: now,
  };
}

export function createMarginNote(
  db: Database.Database,
  highlightId: string,
  content: string,
): MarginNoteRecord | { error: string } {
  // Verify highlight exists and get document_id for touch
  const highlight = db
    .prepare("SELECT id, document_id FROM highlights WHERE id = ?")
    .get(highlightId) as { id: string; document_id: string } | undefined;

  if (!highlight) {
    return { error: `Highlight not found: ${highlightId}` };
  }

  const id = randomUUID();
  const now = nowMillis();

  db.prepare(
    `INSERT INTO margin_notes (id, highlight_id, content, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?)`,
  ).run(id, highlightId, content, now, now);

  touchDocument(db, highlight.document_id);

  return {
    id,
    highlight_id: highlightId,
    content,
    created_at: now,
    updated_at: now,
  };
}

export function deleteHighlight(
  db: Database.Database,
  highlightId: string,
): { success: true } | { error: string } {
  // Get document_id before delete for touch
  const highlight = db
    .prepare("SELECT document_id FROM highlights WHERE id = ?")
    .get(highlightId) as { document_id: string } | undefined;

  if (!highlight) {
    return { error: `Highlight not found: ${highlightId}` };
  }

  // Cascading delete removes attached margin notes
  db.prepare("DELETE FROM highlights WHERE id = ?").run(highlightId);
  touchDocument(db, highlight.document_id);

  return { success: true };
}

export function updateMarginNote(
  db: Database.Database,
  noteId: string,
  content: string,
): MarginNoteRecord | { error: string } {
  const existing = db
    .prepare(
      `SELECT mn.id, mn.highlight_id, mn.created_at, h.document_id
       FROM margin_notes mn
       JOIN highlights h ON mn.highlight_id = h.id
       WHERE mn.id = ?`,
    )
    .get(noteId) as { id: string; highlight_id: string; created_at: number; document_id: string } | undefined;

  if (!existing) {
    return { error: `Margin note not found: ${noteId}` };
  }

  const now = nowMillis();
  db.prepare("UPDATE margin_notes SET content = ?, updated_at = ? WHERE id = ?").run(
    content,
    now,
    noteId,
  );

  touchDocument(db, existing.document_id);

  return {
    id: noteId,
    highlight_id: existing.highlight_id,
    content,
    created_at: existing.created_at,
    updated_at: now,
  };
}

export function deleteMarginNote(
  db: Database.Database,
  noteId: string,
): { success: true } | { error: string } {
  const existing = db
    .prepare(
      `SELECT mn.id, h.document_id
       FROM margin_notes mn
       JOIN highlights h ON mn.highlight_id = h.id
       WHERE mn.id = ?`,
    )
    .get(noteId) as { id: string; document_id: string } | undefined;

  if (!existing) {
    return { error: `Margin note not found: ${noteId}` };
  }

  db.prepare("DELETE FROM margin_notes WHERE id = ?").run(noteId);
  touchDocument(db, existing.document_id);

  return { success: true };
}

export function updateHighlightColor(
  db: Database.Database,
  highlightId: string,
  color: string,
): { success: true } | { error: string } {
  if (!ALLOWED_COLORS.includes(color as (typeof ALLOWED_COLORS)[number])) {
    return {
      error: `Invalid color "${color}". Allowed: ${ALLOWED_COLORS.join(", ")}`,
    };
  }

  const highlight = db
    .prepare("SELECT document_id FROM highlights WHERE id = ?")
    .get(highlightId) as { document_id: string } | undefined;

  if (!highlight) {
    return { error: `Highlight not found: ${highlightId}` };
  }

  db.prepare("UPDATE highlights SET color = ?, updated_at = ? WHERE id = ?").run(
    color,
    nowMillis(),
    highlightId,
  );

  touchDocument(db, highlight.document_id);

  return { success: true };
}

export interface TextLocation {
  from_pos: number;
  to_pos: number;
  text_content: string;
  prefix_context: string;
  suffix_context: string;
}

export function findTextInDocument(
  db: Database.Database,
  documentId: string,
  textToHighlight: string,
): TextLocation | { error: string } {
  const result = readDocument(db, documentId);
  if ("error" in result) {
    return result;
  }

  const content = result.content;
  const idx = content.indexOf(textToHighlight);
  if (idx === -1) {
    return { error: `Text not found in document: "${textToHighlight.slice(0, 80)}"` };
  }

  const fromPos = idx;
  const toPos = idx + textToHighlight.length;
  const prefixStart = Math.max(0, fromPos - 50);
  const suffixEnd = Math.min(content.length, toPos + 50);

  return {
    from_pos: fromPos,
    to_pos: toPos,
    text_content: textToHighlight,
    prefix_context: content.slice(prefixStart, fromPos),
    suffix_context: content.slice(toPos, suffixEnd),
  };
}

export interface HighlightByTextResult {
  highlight: HighlightRecord;
  note?: MarginNoteRecord;
}

export function highlightByText(
  db: Database.Database,
  params: {
    document_id: string;
    text_to_highlight: string;
    color: string;
    note?: string;
  },
): HighlightByTextResult | { error: string } {
  const location = findTextInDocument(db, params.document_id, params.text_to_highlight);
  if ("error" in location) {
    return location;
  }

  const highlight = createHighlight(db, {
    document_id: params.document_id,
    color: params.color,
    text_content: location.text_content,
    from_pos: location.from_pos,
    to_pos: location.to_pos,
    prefix_context: location.prefix_context,
    suffix_context: location.suffix_context,
  });
  if ("error" in highlight) {
    return highlight;
  }

  const result: HighlightByTextResult = { highlight };

  if (params.note) {
    const noteResult = createMarginNote(db, highlight.id, params.note);
    if ("error" in noteResult) {
      return noteResult;
    }
    result.note = noteResult;
  }

  return result;
}
